discovery-provider/src/queries/get_tips.py

```diff
@@ -0,0 +1,288 @@
+import logging
+from datetime import datetime
+from typing import List, Tuple, TypedDict, Union, cast
+
+from sqlalchemy import func, or_
+from sqlalchemy.orm import Query, aliased
+from sqlalchemy.orm.session import Session
+from src.models import AggregateUser, AggregateUserTips, Follow, User, UserTip
+from src.queries.get_unpopulated_users import get_unpopulated_users
+from src.queries.query_helpers import paginate_query, populate_user_metadata
+from src.utils.db_session import get_db_read_replica
+
+logger = logging.getLogger(__name__)
+
+
+class GetTipsArgs(TypedDict):
+    limit: int
+    offset: int
+    user_id: int
+    receiver_min_followers: int
+    receiver_is_verified: bool
+    current_user_follows: str
+    unique_by: str
+    min_slot: int
+    tx_signatures: List[int]
+
+
+class TipResult(TypedDict):
+    amount: int
+    sender: str
+    receiver: str
+    slot: int
+    created_at: datetime
+    followee_supporters: List[str]
+    tx_signature: str
+
+
+# Example of query with inputs:
+# limit=100, offset=0, user_id=1, current_user_follows=sender_or_receiver
+# ---------------------------------------------------------
+# WITH followees AS
+# (
+#     SELECT
+#         follows.followee_user_id AS followee_user_id
+#     FROM
+#         follows
+#     WHERE
+#         follows.is_current = true
+#         AND follows.is_delete = false
+#         AND follows.follower_user_id = 1
+# )
+# ,
+# tips AS
+# (
+#     SELECT
+#         user_tips.signature AS signature,
+#         user_tips.slot AS slot,
+#         user_tips.sender_user_id AS sender_user_id,
+#         user_tips.receiver_user_id AS receiver_user_id,
+#         user_tips.amount AS amount,
+#         user_tips.created_at AS created_at
+#     FROM
+#         user_tips
+#         LEFT OUTER JOIN
+#             followees AS followees_for_receiver
+#             ON user_tips.receiver_user_id = followees_for_receiver.followee_user_id
+#         LEFT OUTER JOIN
+#             followees AS followees_for_sender
+#             ON user_tips.sender_user_id = followees_for_sender.followee_user_id
+#     WHERE
+#         followees_for_receiver.followee_user_id IS NOT NULL
+#         OR followees_for_sender.followee_user_id IS NOT NULL
+#     ORDER BY
+#         user_tips.slot DESC LIMIT 100 OFFSET 0
+# )
+# ,
+# followee_tippers AS
+# (
+#     SELECT
+#         aggregate_user_tips.sender_user_id AS sender_user_id,
+#         aggregate_user_tips.receiver_user_id AS receiver_user_id,
+#         followees_for_aggregate.followee_user_id AS followee_user_id
+#     FROM
+#         followees AS followees_for_aggregate
```

Comment:
is it possible to use tips since followees isn't limited?

Reply:
You mean swap the `FROM` and `JOIN` tables?

Reply:
oh i meant the tips CTE. it looks like all you need is the followee_user_id and that was limited in the tips CTE, right?

Reply:
ooh hmmm that's a good point, we could filter to only show tippers you folllow that are in the same tips query? I think that changes the output but might be a worthwhile change?

Reply:
oh yeah i guess it would be a subset. maybe worth it if there's a big impact on perf.